providers/gitlab.go

```diff
@@ -34,28 +39,89 @@ func NewGitLabProvider(p *ProviderData) *GitLabProvider {
 		p.ValidateURL = &url.URL{
 			Scheme: "https",
 			Host:   "gitlab.com",
-			Path:   "/api/v4/user",
+			Path:   "/api/v4",
 		}
 	}
-	if p.Scope == "" {
-		p.Scope = "read_user"
-	}
+	p.Scope += " read_user"
 	return &GitLabProvider{ProviderData: p}
 }
 
 // GetEmailAddress returns the Account email address
 func (p *GitLabProvider) GetEmailAddress(s *SessionState) (string, error) {
+	if p.Group != "" {
+		groupsJSON, groupsJSONError := data(p.ValidateURL.String()+"/groups?access_token="+s.AccessToken, s)
+		if groupsJSONError != nil {
+			return "", groupsJSONError
+		}
+		groupsValid := false
+		for i := range groupsJSON.MustArray() {
+			group, groupJSONError := groupsJSON.GetIndex(i).Get("name").String()
+			if groupJSONError == nil {
+				groupsValid = groupsValid || (strings.ToLower(group) == p.Group)
+			}
+		}
+		if !groupsValid {
+			return "", fmt.Errorf("user has no access to group %s", p.Group)
+		}
+	}
+
+	primaryUserEmailJSON, primaryUserEmailJSONError := data(p.ValidateURL.String()+"/user?access_token="+s.AccessToken, s)
+	if primaryUserEmailJSONError != nil {
+		return "", primaryUserEmailJSONError
+	}
+	primaryUserEmail, primaryUserEmailError := primaryUserEmailJSON.Get("email").String()
```

Comment:
Is it possible to check that emails are verified? Should we be doing so here?

Reply:
I fear not. I checked the API but I only get all email addresses. There is no information if it is verified or not. Even worse, I also get all unverified email addresses as a result of the request. I will file an issue at gitlab.

Reply:
https://gitlab.com/gitlab-org/gitlab-ce/issues/56779

Reply:
It's not mine, but I gave a `like` to it.